### PR TITLE
Add Horizontal Padding to Constrained Chip Label Calculations

### LINKED ISSUE #11523: Chip does not constrain size of children
## Steps to Reproduce
I'm trying to utilize the Text widget's overflow properties when I put it into a chip.
However when I do the following the chip is not constraining the size of Text and results in the Text widget being wider than the available space*:
```
new Container(
  constraints: new BoxConstraints(maxWidth: 150.0),
  child: new Chip(
    label: new Text(
      'Some text that exceeds the width of parent',
      overflow: TextOverflow.ellipsis),
    onDeleted: myDeleteFunction
)
```
It produces the following: 
screen shot 2017-08-04 at 3 06 55 pm

*putting 150.0 for illustrative purposes only, my main use case is this chip in a wrap

**Current work around**
Instead I need to do something like the following - where 38 represents the extra space on the "right" including the "x" icon since I always have a onDeleted function. This, unfortunately, requires me to hard-code the 38.
```dart
new Chip(
    label: new Container(
        constraints: new BoxConstraints(
            maxWidth: 150.0 - 38.0),
        child: new Text(email,
            overflow: TextOverflow.ellipsis)),
    onDeleted: myDeleteFunction,
)
```
Which works and gives me this:
screen shot 2017-08-04 at 3 07 08 pm

## Request
**Would it be possible for the label in Chip to be wrapped with an Expanded?**
Something like this:
```dart
    children.add(new Expanded(
      child: new DefaultTextStyle(
        style: _kLabelStyle,
        child: label
      )
    ));
```

### LINKED ISSUE #31790: InputChip delete button is misaligned
## Steps to Reproduce

When using e.g. InputChips with onDeleted, the Delete icon is placed outside the chip if the text width is larger than the allowed width for the chip.

Screenshot_20190429-145705

```dart
import 'package:flutter/material.dart';

void main() {
  return runApp(TestChips());
}

class TestChips extends StatelessWidget {
  final String lorem =
      "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam "
      "nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, "
      "sed diam voluptua. At vero eos et accusam et justo duo dolores et ea "
      "rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem "
      "ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing "
      "elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna "
      "aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo "
      "dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus "
      "est Lorem ipsum dolor sit amet.";

  @override
  Widget build(BuildContext context) {
    List<Widget> chips = [];

    for (var i = 0; i < 12; i++) {
      chips.add(InputChip(
        label: Text(lorem.substring(0, i * 2)),
        onDeleted: () => print("delete"),
      ));
    }

    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text("Test Chips"),
        ),
        body: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Container(
            width: 150,
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: chips,
            ),
          ),
        ),
      ),
    );
  }
}
```

## Logs

```
flutter analyze
No issues found! (ran in 3.3s)
```

```
flutter doctor -v
[✓] Flutter (Channel stable, v1.2.1, on Linux, locale de_DE.utf8)
    • Flutter version 1.2.1 at /home/xxx/dev_tools/flutter
    • Framework revision 8661d8aecd (vor 2 Monaten), 2019-02-14 19:19:53 -0800
    • Engine revision 3757390fa4
    • Dart version 2.1.2 (build 2.1.2-dev.0.0 0a7dcf17eb)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /home/xxx/dev_tools/android_sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /home/xxx/dev_tools/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b16-5323222)
    • All Android licenses accepted.

[✓] Android Studio (version 3.4)
    • Android Studio at /home/xxx/dev_tools/android-studio
    • Flutter plugin version 35.0.1
    • Dart plugin version 183.5901
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b16-5323222)

[✓] Connected device (1 available)
    • HTC U11 life • HT7AY1C00279 • android-arm64 • Android 9 (API 28)

• No issues found!

```

### PR DESCRIPTION
**Pull request by shihaohong**
## Description

Currently, `InputChips` do not account for two things:
1) `_layoutLabel` now returns a `Size` that is properly constrained if there is a finite `constraint.maxWidth`.
2) The constraint calculations do not account for horizontal padding for the chip's shape when it has a finite constraint. This change incorporates it in layout calculations.

I also found that existing tests in `chip_test.dart` that utilize the `_testConstrainedLabel` have a brittle way of determining that a `Chip` and its avatar, label or delete icon is properly constrained by its parent, as it just checks to see if the label and chip's width and height are less than or equal to it's parent's width and height, since this does not account for:
1) whether or not they are constrained positionally,
2) whether or not having a combination of avatar, label and delete icon will cause overflow.

After some investigation, a cleaner fix that lays out the above will be complex, so this PR resolves the immediate issue of constraining the label and delete icon to a match its parent. I will file issues to track this need.

TODO:
- [x] Create an issue regarding updating the tests and layout (https://github.com/flutter/flutter/issues/31914)

## Related Issues

Fixes https://github.com/flutter/flutter/issues/31790
Fixes https://github.com/flutter/flutter/issues/11523

## Tests

- Regression test validating that all child widgets are constrained to the width of the Chip's parent widget

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/chip.dart [hunk 1]
```diff
@@ -2188,29 +2188,40 @@ class _RenderChip extends RenderBox {
     // Now that we know the label height and the width of the icons, we can
     // determine how much to shrink the width constraints for the "real" layout.
     if (constraints.maxWidth.isFinite) {
+      final double maxWidth = math.max(
+        0.0,
+        constraints.maxWidth
+        - iconSizes
+        - theme.labelPadding.horizontal
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip.dart]: Is the labelPadding applied unconditionally, or only if there's an avatar and a deleteIcon (in that case would it be applied twice?

File: packages/flutter/test/material/chip_test.dart [hunk 2]
```diff
@@ -248,6 +248,97 @@ void main() {
...
+    Rect deleteIconRect;
+    const String text = 'very long text that be clipped';
+
+    await tester.pumpWidget(chipBuilder(text));
+
+    chipRect = tester.getRect(find.byType(Chip));
+    labelRect = tester.getRect(
+      find.text(text),
+    );
+    expect(chipRect.contains(labelRect.topLeft), true);
```
Comment (HansMuller) [inline packages/flutter/test/material/chip_test.dart]: Might be a little simpler to write a local chipRectContains() function

**Review by HansMuller — COMMENTED** [review]
Looks good, just wanted sort out one fine point about padding.

File: packages/flutter/lib/src/material/chip.dart (hunk 1 above)
Comment (shihaohong) [reply packages/flutter/lib/src/material/chip.dart]: It should be applied unconditionally according to https://material.io/design/components/chips.html#specs

**Review by HansMuller — APPROVED** [review]
LGTM

**Pull request merged**